synapse/visibility.py

```diff
@@ -565,29 +575,40 @@ async def filter_events_for_server(
     storage: StorageControllers,
     target_server_name: str,
     local_server_name: str,
-    events: List[EventBase],
+    events: Sequence[EventBase],
     redact: bool = True,
-    check_history_visibility_only: bool = False,
+    filter_out_erased_senders: bool = True,
 ) -> List[EventBase]:
-    """Filter a list of events based on whether given server is allowed to
+    """Filter a list of events based on whether the target server is allowed to
     see them.
 
+    For a fully stated room, the target server is allowed to see an event E if:
+      - the state at E has world readable or shared history vis, OR
+      - the state at E says that the target server is in the room.
+
+    For a partially stated room, the target server is allowed to see E if:
+      - E was created by this homeserver, AND:
+          - the partial state at E has world readable or shared history vis, OR
+          - the partial state at E says that the target server is in the room.
```

Comment:
The fully stated section is my reading of the original code.

The partially stated section is what I _think_ we are trying to do in the original code, and what I think we should be doing.